### PR TITLE
Optimize Enum.random/1, Enum.take_random/2 for ranges

### PR DESCRIPTION
**Pull request by eksperimental**
Private function Enum.random_index/2 has been replaced with Enum.random_integer/2
and the deterministic examples had to be updated.

### CODE REVIEW

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -2494,9 +2499,17 @@ defmodule Enum do
   defp enum_to_string(entry) when is_binary(entry), do: entry
   defp enum_to_string(entry), do: String.Chars.to_string(entry)
 
-  defp random_index(n) do
-    :rand.uniform(n + 1) - 1
-  end
+  @spec random_integer(integer, integer) :: integer
+  defp random_integer(lower_limit \\ 0, upper_limit)
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: I usually don't like using default arguments for private functions. And this one is making the first argument optional, which is even less clearer. What if we remove the default arg and pass it explicitly above? :)

**Comment by josevalim** [comment]
@eksperimental beautiful, thank you!

File: lib/elixir/lib/enum.ex [hunk 2]
```diff
@@ -1542,18 +1542,25 @@ defmodule Enum do
   ## Examples
 
       # Although not necessary, let's seed the random algorithm
-      iex> :rand.seed(:exsplus, {1, 2, 3})
-      iex> Enum.random([1, 2, 3])
-      2
+      iex> :rand.seed(:exsplus, {9, 3, 4})
```
Comment (whatyouhide) [inline lib/elixir/lib/enum.ex]: Why this change? Users will never seed this by hand (hopefully), so I think it may even be better to use `{1, 2, 3}` to make that even clearer?

**Comment by eksperimental** [comment]
changes updated

File: lib/elixir/lib/enum.ex (hunk 1 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: fixed

File: lib/elixir/lib/enum.ex (hunk 2 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: fixed

File: lib/elixir/lib/enum.ex [hunk 3]
```diff
@@ -2494,9 +2499,17 @@ defmodule Enum do
   defp enum_to_string(entry) when is_binary(entry), do: entry
   defp enum_to_string(entry), do: String.Chars.to_string(entry)
 
-  defp random_index(n) do
-    :rand.uniform(n + 1) - 1
-  end
+  @spec random_integer(integer, integer) :: integer
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: I think we don't do `@spec` for private functions.

File: lib/elixir/lib/enum.ex [hunk 4]
```diff
@@ -2494,9 +2499,17 @@ defmodule Enum do
...
+  defp random_integer(lower_limit, upper_limit)
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: This clause is not needed as function is private.

File: lib/elixir/lib/enum.ex (hunk 3 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: We actually do in a few places:
```
$ ag "@spec [^\n]+\s+defp"

lib/elixir/lib/dict.ex
364:  @spec unsupported_dict(t) :: no_return
365:  defp unsupported_dict(dict) do

lib/elixir/lib/macro.ex
79:  @spec binary_op_props(atom) :: {:left | :right, precedence :: integer}
80:  defp binary_op_props(o) do

lib/elixir/lib/inspect/algebra.ex
537:  @spec fits?(integer, [{integer, mode, t}]) :: boolean
538:  defp fits?(w, _) when w < 0,                      do: false
549:  @spec format(integer | :infinity, integer, [{integer, mode, t}]) :: [binary]
550:  defp format(_, _, []),                                do: []
```

I remember I asked this a while ago to José, and he mentioned it's OK since it can useful for Dialyzer.

File: lib/elixir/lib/enum.ex (hunk 4 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: agreed

**Comment by eksperimental** [comment]
does anybody know how can I represent a typespec for
`@spec random_integer(integer, integer) :: integer`
that the return value would be within `lower_limit..upper_limit`

/cc @ericmj

File: lib/elixir/lib/enum.ex (hunk 3 above)
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: @eksperimental those are rather the exception than the rule. :) Given the current spec is over generic anyway, I would remove it. :)

File: lib/elixir/lib/enum.ex [hunk 5]
```diff
@@ -2259,14 +2266,12 @@ defmodule Enum do
     do: []
   def take_random(first..first, count) when is_integer(count) and count >= 1,
     do: [first]
-  def take_random(first..last, 1) when first > last,
-    do: take_random(last..first, 1)
-  def take_random(first..last, 1),
-    do: [random_index(last - first) + first]
+  def take_random(_.._ = range, 1),
+    do: [random(range)]
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: This changes the semantic of the code, doesn't it? It means that, the following will no longer be true when starting from the same seed:

```elixir
:rand.seed(:exsplus, {101, 102, 103})
one = Enum.take_random(1..100, 1)

:rand.seed(:exsplus, {101, 102, 103})
two = Enum.take_random(1..100, 2)

hd(one) == hd(two)
```

Comment (josevalim) [inline lib/elixir/lib/enum.ex]: We need to either change behaviour for all ranges or stop special casing `1` inside `take_random`. I think it should be simpler to change behaviour for all ranges. Note this problem does not exist for Enum.random (because it is always `1` and we don't provide any guarantee between `random` and `take_random`).

File: lib/elixir/lib/enum.ex [hunk 6]
```diff
@@ -1542,18 +1542,25 @@ defmodule Enum do
...
+
   def random(enumerable) do
-    case take_random(enumerable, 1) do
-      [] -> raise Enum.EmptyError
-      [e] -> e
+    case count(enumerable) do
+      0 ->
+        raise Enum.EmptyError
+      count ->
+        at(enumerable, random_integer(0, count - 1))
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: This is nice. I am assuming this is reasonably faster too. Have you measured?

File: lib/elixir/lib/enum.ex (hunk 5 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: sorry jose, but that wasn't true before this change

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: Ok.. i see what you mean, so should we remove this clause bellow?
```elixir
  defp random_integer(limit, limit) when is_integer(limit),
    do: limit
```
and also get rid of the optimizations

```elixir
  def take_random(_enumerable, 0),
    do: []
  def take_random(first..first, count) when is_integer(count) and count >= 1,
    do: [first]
```

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: the code you mentioned above worked, for the 1.2 branch,
in 1.3 there are two commits of my own that changed the semantics, which are: 76d175a0fb44e460c1699e7eaed6a8acc7113cf5 and 2d707e25f70d66e39eff8ff0a25c9a1dc5306019

let me know how should we proceed.
I'm so sorry about this, didn't think of this at all.

We should definitely add your example as a test, and also one dealing with an empty enumerable

Comment (josevalim) [inline lib/elixir/lib/enum.ex]: It is OK to keep your optimization as long as we always apply them to ranges regardless of the `take_random` count.

File: lib/elixir/lib/enum.ex (hunk 6 above)
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: @josevalim I've measured this change for ranges, maps, and lists. It is reasonably faster indeed (5-8 times depending on size and type).

File: lib/elixir/lib/enum.ex [hunk 7]
```diff
@@ -2494,9 +2499,14 @@ defmodule Enum do
...
 
-  defp random_index(n) do
-    :rand.uniform(n + 1) - 1
-  end
+  defp random_integer(limit, limit) when is_integer(limit),
+    do: limit
+  defp random_integer(lower_limit, upper_limit) when is_integer(lower_limit) and is_integer(upper_limit),
+    do: do_random_integer(lower_limit, upper_limit)
+  defp do_random_integer(lower_limit, upper_limit) when upper_limit > lower_limit,
+    do: lower_limit + :rand.uniform(upper_limit - lower_limit + 1) - 1
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: I think this expression too complex to be used with `do:` style.

File: lib/elixir/lib/enum.ex [hunk 8]
```diff
@@ -2494,9 +2499,14 @@ defmodule Enum do
...
   defp enum_to_string(entry), do: String.Chars.to_string(entry)
 
-  defp random_index(n) do
-    :rand.uniform(n + 1) - 1
-  end
+  defp random_integer(limit, limit) when is_integer(limit),
+    do: limit
+  defp random_integer(lower_limit, upper_limit) when is_integer(lower_limit) and is_integer(upper_limit),
+    do: do_random_integer(lower_limit, upper_limit)
+  defp do_random_integer(lower_limit, upper_limit) when upper_limit > lower_limit,
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: Seems we can ditch `do_random_integer` by defining `random_integer` as follows:
```elixir
  defp random_integer(limit, limit) when is_integer(limit),
    limit
  end

  defp random_integer(lower_limit, upper_limit) when upper_limit < lower_limit,
    random_integer(upper_limit, lower_limit)
  end

  defp random_integer(lower_limit, upper_limit) when is_integer(lower_limit) and is_integer(upper_limit) do
    lower_limit + :rand.uniform(upper_limit - lower_limit + 1) - 1
  end
```

File: lib/elixir/lib/enum.ex (hunk 6 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: yes, it is faster in every single instance (i have excluded ranges since they are not part of the code in question),
bench file: https://gist.github.com/eksperimental/d9d98c1f4e47ecbcb19a6affe872dff4

```
$ mix bench -d 3
Settings:
  duration:      3.0 s

## EnumRandomBench
benchmark name                         iterations   average time 
Elixir.EnumRandomNew: list_empty         10000000   0.42 µs/op
Elixir.EnumRandomNew: map_empty          10000000   0.51 µs/op
Elixir.EnumRandomOriginal: list_empty    10000000   0.63 µs/op
Elixir.EnumRandomOriginal: map_empty      5000000   1.08 µs/op
Elixir.EnumRandomNew: list                1000000   8.03 µs/op
Elixir.EnumRandomNew: map                  100000   62.46 µs/op
Elixir.EnumRandomOriginal: list              5000   1030.92 µs/op
Elixir.EnumRandomOriginal: map               5000   1079.91 µs/op
Elixir.EnumRandomNew: list_big               5000   1342.06 µs/op
Elixir.EnumRandomNew: map_big                 500   13701.73 µs/op
Elixir.EnumRandomOriginal: list_big            50   223223.72 µs/op
Elixir.EnumRandomOriginal: map_big             50   251013.74 µs/op
``` 

see the results for the big maps and lists, the difference is HUGE, and the they have 200,000 elements

File: lib/elixir/lib/enum.ex [hunk 9]
```diff
@@ -2494,8 +2499,18 @@ defmodule Enum do
...
   defp enum_to_string(entry), do: String.Chars.to_string(entry)
 
-  defp random_index(n) do
-    :rand.uniform(n + 1) - 1
+  defp random_integer(limit, limit) when is_integer(limit) do
+    limit
+  end
+
+  defp random_integer(lower_limit, upper_limit)
+      when is_integer(lower_limit) and is_integer(upper_limit) and upper_limit < lower_limit do
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: It's possible to skip `is_integer(lower_limit) and is_integer(upper_limit)` check here since next clause won't match for non-integers anyway. :bowtie:

**Comment by ericmj** [comment]
> does anybody know how can I represent a typespec for
@spec random_integer(integer, integer) :: integer
that the return value would be within lower_limit..upper_limit

```elixir
@spec random_integer(lower, upper) :: lower..upper
when lower: integer, upper: integer
```

But we usually don't typespec private functions.

File: lib/elixir/lib/enum.ex (hunk 6 above)
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: Btw, this will traverse the enumerable twice, which is something we typically do not do. It is fine for now, I will address this after merged.

Comment (michalmuskala) [inline lib/elixir/lib/enum.ex]: Wouldn't this break for streams that can't be enumerated multiple times, like files?

File: lib/elixir/lib/enum.ex (hunk 5 above)
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: The new implementation should just call random_integer(...) the amount of times desired. :) Something like:

```elixir
def take_random(_.._ = range, n) do
  take_random_range(range, n)
end

defp take_random_range(_range, 0), do: 0
defp take_random_range(range, n), do: [random(range) | take_random_range(range, n - 1)]
```

File: lib/elixir/lib/enum.ex (hunk 6 above)
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: It will, I will address it after merged. :)

**Comment by eksperimental** [comment]
@ericmj that's exactly what I was looking for! thank you

File: lib/elixir/lib/enum.ex (hunk 5 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: thank you José, but that wouldn't work. take_random returns unique values, and it can never return more element than the original enumerable had.
that would be the implementation for ranges for `Enum.random/2` which is a new function I was planning on propose after this PR. To contrast with `Enum.take_random/2` in the sense that this function may return repeated values, and it will return a list of the exact number of count elements given.

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: sorry, not unique, but if the given enumerable has unique values, it will never return repeated values

Comment (josevalim) [inline lib/elixir/lib/enum.ex]: Oh, good catch. So let's revert these changes for now and add a test. We
can discuss Enum.random/2 separately.

On Monday, July 4, 2016, Eksperimental <notifications@github.com> wrote:

> In lib/elixir/lib/enum.ex
> <https://github.com/elixir-lang/elixir/pull/4927#discussion_r69400684>:
>
> > @@ -2259,14 +2266,12 @@ defmodule Enum do
> >      do: []
> >    def take_random(first..first, count) when is_integer(count) and count >= 1,
> >      do: [first]
> > -  def take_random(first..last, 1) when first > last,
> > -    do: take_random(last..first, 1)
> > -  def take_random(first..last, 1),
> > -    do: [random_index(last - first) + first]
> > +  def take_random(_.._ = range, 1),
> > +    do: [random(range)]
>
> sorry, not unique, but if the given enumerable has unique values, it will
> never return repeated values
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/4927/files/8572da663e212f709689d504fa954f315bedd7bd#r69400684>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAlbij6qxPBk9u4RzINKtNB69MVOVhTks5qSF5LgaJpZM4JD0MS>
> .
>


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

File: lib/elixir/lib/enum.ex [hunk 10]
```diff
@@ -2494,8 +2499,18 @@ defmodule Enum do
...
+    random_integer(upper_limit, lower_limit)
+  end
+
+  defp random_integer(lower_limit, upper_limit)
+      when is_integer(lower_limit) and is_integer(upper_limit) do
```
Comment (ericmj) [inline lib/elixir/lib/enum.ex]: What's our convention on guard clause indentation? Don't think I have seen 4 spaces before.

Comment (whatyouhide) [inline lib/elixir/lib/enum.ex]: It's 4 spaces in Elixir source code, yep :)

**Comment by eksperimental** [comment]
@josevalim we are for a final review.
the only issue left to be addressed is the one about traversing the enumerable twice https://github.com/elixir-lang/elixir/pull/4927/files#r69396589

File: lib/elixir/lib/enum.ex (hunk 6 above)
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: @eksperimental given everything is handled, I think we should fix this by doing something like this:

```elixir
case Enumerable.count(data) do
  {:ok, 0} ->
    raise Enum.EmptyError
  {:ok, count} ->
    at(enumerable, random_integer(0, count - 1))
  {:error, _} ->
    case take_random(enumerable, 1) do
      []  -> raise Enum.EmptyError
      [e] -> e
    end
end
```

What do you think? Can you please amend accordingly if you agree?

Thank you! :heart:

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: sure I can do that.
just out of curiosity, do you have an example that would not work with the previous code, and it will with this new one.

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: @josevalim that's pretty ingenious how to not to traverse enumerable twice.
code update, we are good to merge

Comment (josevalim) [inline lib/elixir/lib/enum.ex]: You can use something like Stream.resource:

```elixir
        Stream.resource(fn -> Process.put(:stream_started, true) && raise "started twice"; 0 end,
                        fn 0 -> {[0], 1}; 1 -> {:halt, 1} end,
                        fn _ -> :ok end)
```

Calling `random` on it should return `0` and not raise the inner started twice error.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by eksperimental** [comment]
@josevalim should we backport by removing these changes realted to ranges with count =1, that changed the semantics?
https://github.com/elixir-lang/elixir/commit/b6245074614b241f0df2c0eedaf981412f67f71c#diff-6881431a92cd4e3ea0de82bf2338f8eaL2262

initially introduced in 095321451743ea000201b5f0ff3456a061ed62ba

**Comment by josevalim** [comment]
I won't backport because I believe we should not change the random result for a given seed in a tiny release.

**Comment by eksperimental** [comment]
great then. just wanted to make sure we are not missing anything.
